### PR TITLE
Refactor GA4 code

### PR DESCRIPTION
**Pull request by andysellick**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What
Refactor some of the GA4 code in `finder-frontend`, based on @floehopper's previous work in https://github.com/alphagov/finder-frontend/pull/3166. This moves the logic for generating indexes for GA4 tracking on search elements (the option select and expander components) from the JS into Ruby.

Now depends upon this change in the components gem: https://github.com/alphagov/govuk_publishing_components/pull/3750

## Why
We want our code to be clean and readable.

## Visual changes
None.

Trello card: https://trello.com/c/OdYXlYqQ/673-finder-frontend-code-optimisations

### CODE REVIEW

**Review by AshGDS — CHANGES_REQUESTED** [review]
Looks good, though a small is change needed :+1:

File: app/views/finders/_radio_facet.html.erb
Comment (andysellick) [reply app/views/finders/_radio_facet.html.erb]: Thanks @AshGDS, have restored.

**Review by AshGDS — DISMISSED** [review]
LGTM :+1:

**Comment by andysellick** [comment]
Thanks @AshGDS. I'm currently looking at expanding this PR further, which is going to involve some changes in the gem first, so I'm going to pause on this for the moment (you might want to withdrawn your approval).

**Comment by andysellick** [comment]
@AshGDS have added some more changes to this PR. This should now fully remove the JS for event listening and setting indexes on the option select and expander component buttons. It's related to [this change in the gem](https://github.com/alphagov/govuk_publishing_components/pull/3750) for the option select, so will need to be tested together and I'll figure out how to deploy these two changes simultaneously.

Haven't updated the docs yet, that's another thing to look at, but should be enough to review for now.

**Comment by AshGDS** [comment]
@andysellick Looks good to me :+1:

**Comment by andysellick** [comment]
@AshGDS I've bumped the gem in this PR to include the needed changes to the option select component. If you have time can you give this a final once over to check I've not overlooked anything?

I haven't looked at documentation yet, I've just realised... I might do that in a separate PR for the sake of simplicity.

File: app/assets/javascripts/components/expander.js
Comment (AshGDS) [inline app/assets/javascripts/components/expander.js:74]: Does this need to be in a `try/catch`?

**Review by AshGDS — APPROVED** [review]
Looks good :+1:

File: app/assets/javascripts/components/expander.js
Comment (andysellick) [reply app/assets/javascripts/components/expander.js]: Good shout, will match the same thing in the option select. Have added.

File: app/assets/javascripts/components/expander.js
Comment (AshGDS) [inline app/assets/javascripts/components/expander.js:77]: I just noticed something with this regex which also applies to the one in `govuk_publishing_components`

Would `/i` be irrelevant here as it's capturing a symbol rather than a character, so setting it to insensitive isn't necessary?

Also, should it have the `/g` flag instead so that it replaces every instance of an underscore in the key? With the current regex, I think it only works if the key only has one underscore?

**Review by AshGDS — COMMENTED** [review]
Thanks for adding the `try/catch`. I may have noticed something regarding the regex

File: app/assets/javascripts/components/expander.js
Comment (andysellick) [reply app/assets/javascripts/components/expander.js]: That's really helpful, thanks @AshGDS. Yes I think you're right, although for readability I've gone with the `replaceAll` function. I've also updated the tests to reflect this.

Comment (andysellick) [reply app/assets/javascripts/components/expander.js]: Once I'm happy with this I'll make the same change on the option select component in the gem as well.

Comment (AshGDS) [reply app/assets/javascripts/components/expander.js]: Thanks @andysellick - just to confirm, are we dropping IE11 support? If I recall correctly `replaceAll` doesn't work in IE11 (and it seems to not be working when I check in IE11 via BrowserStack). If we're not supporting IE11 anymore then happy to approve the change :+1:

Comment (andysellick) [reply app/assets/javascripts/components/expander.js]: Ah thanks for spotting that, hadn't realised. Have reverted it back to `replace`.

**Comment by andysellick** [comment]
@AshGDS going to hold off on merging this until tomorrow now. Feel free to take your time if you want to look over it again - it's already proved useful!

**Review by AshGDS — CHANGES_REQUESTED** [review]
Thanks @andysellick :+1: Looks like there's one small thing to fix in the test

**Review by AshGDS — APPROVED** [review]
LGTM 👍

**Pull request merged**
